File: OOPS_Concepts/CSharp/Abstraction.cs
using System;

public abstract class Payment
{
    public abstract void ProcessPayment();
}

public class CreditCardPayment : Payment
{
    public override void ProcessPayment()
    {
        Console.WriteLine("Processing credit card payment.");
    }
}

public class PayPalPayment : Payment
{
    public override void ProcessPayment()
    {
        Console.WriteLine("Processing PayPal payment.");
    }
}

public interface ILogger
{
    void LogMessage(string message);
}

public class FileLogger : ILogger
{
    public void LogMessage(string message)
    {
        Console.WriteLine($"Logging message to file: {message}");
    }
}

public class ConsoleLogger : ILogger
{
    public void LogMessage(string message)
    {
        Console.WriteLine($"Logging message to console: {message}");
    }
}

public class Program
{
    public static void Main()
    {
        Payment payment1 = new CreditCardPayment();
        Payment payment2 = new PayPalPayment();

        payment1.ProcessPayment();
        payment2.ProcessPayment();

        ILogger fileLogger = new FileLogger();
        ILogger consoleLogger = new ConsoleLogger();

        fileLogger.LogMessage("This is a file log message.");
        consoleLogger.LogMessage("This is a console log message.");
    }
}

File: OOPS_Concepts/Java/Abstraction.java
/**
 * Abstraction in Java
 * 
 * Abstraction is the process of hiding the implementation details and showing only the functionality to the user.
 * It can be achieved through abstract classes and interfaces.
 */

abstract class Payment {
    abstract void processPayment();
}

class CreditCardPayment extends Payment {
    void processPayment() {
        System.out.println("Processing credit card payment.");
    }
}

class PayPalPayment extends Payment {
    void processPayment() {
        System.out.println("Processing PayPal payment.");
    }
}

interface Logger {
    void logMessage(String message);
}

class FileLogger implements Logger {
    public void logMessage(String message) {
        System.out.println("Logging message to file: " + message);
    }
}

class ConsoleLogger implements Logger {
    public void logMessage(String message) {
        System.out.println("Logging message to console: " + message);
    }
}

public class Main {
    public static void main(String[] args) {
        Payment payment1 = new CreditCardPayment();
        Payment payment2 = new PayPalPayment();

        payment1.processPayment();
        payment2.processPayment();

        Logger fileLogger = new FileLogger();
        Logger consoleLogger = new ConsoleLogger();

        fileLogger.logMessage("This is a file log message.");
        consoleLogger.logMessage("This is a console log message.");
    }
}
